ci: Refactor GitHub Actions CI so Node v12, v14, v16 are tested

aloisklink opened the pull request:
I noticed that your GitHub Actions CI didn't catch https://github.com/trezor/trezor-suite/issues/5107, so since I was already familiar with your CI (due to https://github.com/trezor/connect/pull/1050 (may cause merge conflicts)), I thought I'd fix it.

The issue with the previous GitHub Action is that `actions/setup-node@v2` for Node v12 was only called in the `setup` job. Each job in GitHub CI is called on a different VM, so every other job defaulted to Node v16.

Although I could have just copied that setup action to every single job, to avoid code reuse, I've refactored all of the `.github/workflows/test*` files into:
  - `.github/workflows/test.yml`, which is started on a `push`/`pull_request` to the `develop` branch.
  - `.github/workflows/test_with_trezor-user-env.yml`, a workflow that's called from `.github/workflows/test.yml`, which contains all of the reusable code that runs `trezor_user_env` and then the test calls in Node v12, v14, v16, and karma.

Behavior changes:
  - Instead of using a custom call to cache the node_modules folder (not guaranteed to be cross-compatible across Node.js versions), I'm using the [recommended built-in of `actions/setup-node`](https://github.com/actions/setup-node/blob/main/docs/advanced-usage.md#caching-packages-dependencies)
  - Node.JS tests are now run on Node v12, v14, v16 in parallel (maybe fixes https://github.com/trezor/trezor-suite/issues/5111)
  - karma tests are run only on Node v12, since I believe they're browser based, so don't need to be run on multiple Node versions.
  - Instead of using `./tests/run.sh` to pull and start the docker container for `trezor-user-env`, I'm using [GitHub Actions / Containerized Services](https://docs.github.com/en/actions/using-containerized-services/about-service-containers), and pulling the container from [GitHub](https://github.com/trezor/trezor-user-env) instead of GitLab.
    - `./tests/run.sh` is still used to ensure that the container has finished starting.
  - For testing, I've also reverted `@trezor/rollout` to be versions: `"~1.2.0 || ^1.3.2"`, so that Node v12 CI still works.

Potentially fixes https://github.com/trezor/trezor-suite/issues/5111 (Karma tests and `yarn build` are only run on Node v12)
(apologies for stealing something on your todo list, I had already almost completed this before I noticed you already had an issue for it!)

aloisklink committed:
ci: Use Node v12 in GitHub test_methods_nodejs

Previously, since we weren't explicitly calling actions/setup-node@v2
in every step, GitHub Actions was defaulting to using Node v16.

I've also refactored the code to use a reusable "workflow" step,
to avoid code reuse.

aloisklink committed:
ci: Combine NodeJS and karma tests into one file

This also means that karma tests now run using Node v12,
instead of defaulting to Node v16.

aloisklink committed:
ci: Combine Github methods and api into test.yml

This also means that API tests are now made using Node v12,
instead of defaulting to Node v16.

aloisklink committed:
ci: Use GitHub's Actions services runner

Use GitHub's services runner to directly run
the trezor-user-env Docker container.

The `./tests/run.sh` script no longer runs Docker,
however it's still used to ensure that the `trezor-user-env`
Docker container is ready to accept connections.

aloisklink committed:
ci: Run GitHub NodeJs tests with Node 12, 14 & 16

Karma tests are run with Node v12 only, since they're browser-based.

mroz22 commented:
Hi, this contribution looks really nice. I need to get familiar with this approach a bit. Also sorry for not merging #1050 yet. It does not have that high priority at the moment, since we are going to move this project into trezor-sutie repository soon and all of this CI configuration will need some more engineering.

Also I would like to invite @vdovhanych to this.

vdovhanych commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,135 @@
+    - run: yarn test:unit
+
+  build:
+    runs-on: ubuntu-18.04
```
do we still need to use ubuntu 18.04 for the build? Assume the latest would work fine for node12 also, or?

aloisklink replied:
I just used what was already there in the `develop` branch, since I'm unfamiliar with karma/your build tools. https://github.com/trezor/connect/blob/6368fcda5f6abe7f31b5de8c9dea51a5509866a3/.github/workflows/test_api_karma.yml#L11

However, I tried upgrading it to `ubuntu-latest` in 86c0e578ca1989a857504638c1c5244f32cf0817 and it seems to work.

vdovhanych commented on .github/workflows/test_with_trezor-user-env.yml:
```diff
@@ -0,0 +1,84 @@
+
+  test-karma:
+    name: Test Karma
+    runs-on: ubuntu-18.04
```
again, can be latest maybe?

aloisklink replied:
I'm unfamiliar with how `karma` and `xvfb-run` works, but I tried upgrading it to `ubuntu-latest` in 86c0e578ca1989a857504638c1c5244f32cf0817 and it seems to work.

vdovhanych reviewed (commented):
I really like this change I have just a few small things I'm not sure about but in general, it's really good. A more robust approach for sure.

aloisklink committed:
ci: Use ubuntu-latest to build & run karma tests

aloisklink commented:
I've upgraded the `build` and `karma` tests to use `ubuntu-latest` in https://github.com/trezor/connect/pull/1076/commits/86c0e578ca1989a857504638c1c5244f32cf0817, and it seems to work.

Also, there was a bug where the `eth`/`btc`/... tests were testing all the methods, which I've fixed in https://github.com/trezor/connect/pull/1076/commits/0123b020f7153226d2206a3fa26633b3bf3f5e9d
(my fault, I had fixed this on my testing branch, but I think the `git rebase`/`merge` didn't apply it to this branch.

> Also sorry for not merging https://github.com/trezor/connect/pull/1050 yet. It does not have that high priority at the moment, since we are going to move this project into trezor-sutie repository soon and all of this CI configuration will need some more engineering.

No worries :smile: CI changes are generally less important than other PRs!

mroz22 commented:
Hello, I think we are ready for merge. Just please squash fixups and rebase.

mroz22 commented on package.json:
```diff
@@ -65,7 +65,7 @@
         "@trezor/blockchain-link": "^2.1.1",
         "@trezor/connect-common": "^0.0.4",
-        "@trezor/rollout": "^1.3.1",
+        "@trezor/rollout": "~1.2.0 || ^1.3.2",
```
this change should not be needed anymore I think. latest rollout should contain your fix allowing it to run with node 12

aloisklink replied:
Done, I've deleted that commit, after rebasing to `develop`.

vdovhanych commented on .github/workflows/test.yml:
```diff
@@ -0,0 +1,135 @@
+    - run: yarn lint
+    - run: yarn test:unit
+
+  build:
```
We should also build it with node14 and node16 for propper testing in matrix like style
